File: model/src/main/java/io/sparkled/model/render/RenderedStagePropData.java
package io.sparkled.model.render;

import java.util.ArrayList;
import java.util.List;

public class RenderedStagePropData {

    public static final RenderedStagePropData EMPTY = new RenderedStagePropData();

    private final int ledCount;
    private final byte[] data;

    // Don't serialise frames to JSON, as each frame contains a reference to the (very large) data array
    private transient final List<RenderedFrame> frames;

    /**
     * Default constructor required for Gson.
     */
    @SuppressWarnings("unused")
    public RenderedStagePropData() {
        this(0, 0, 0, new byte[]{});
    }

    public RenderedStagePropData(int startFrame, int endFrame, int ledCount, byte[] data) {
        this.ledCount = ledCount;
        this.data = data;

        int frameCount = endFrame - startFrame + 1;
        this.frames = new ArrayList<>(frameCount);

        for (int frameIndex = startFrame; frameIndex <= endFrame; ++frameIndex) {
            this.frames.add(new RenderedFrame(startFrame, frameIndex, ledCount, data));
        }
    }

    public byte[] getData() {
        return data;
    }

    public List<RenderedFrame> getFrames() {
        return frames;
    }

    public int getLedCount() {
        return ledCount;
    }
}

File: model/src/main/kotlin/io/sparkled/model/render/RenderedStagePropData.kt
package io.sparkled.model.render

import com.fasterxml.jackson.annotation.JsonIgnore
import io.sparkled.model.UniqueId
import io.sparkled.model.annotation.GenerateClientType

@GenerateClientType
class RenderedStagePropData(
    startFrame: Int,
    endFrame: Int,
    val ledCount: Int,
    val data: ByteArray,
    @JsonIgnore
    val stagePropRanges: Map<UniqueId, IntRange> = emptyMap(),
) {

    // Don't serialise frames to JSON, as each frame contains a reference to the (very large) data array
    @JsonIgnore
    val frames = (startFrame..endFrame).map {
        RenderedFrame(startFrame, it, ledCount, data)
    }
}
